### ISSUE TITLE
Slow documentation building with images

### ISSUE THREAD

**Issue by sebw** [mkdocs/mkdocs#1331]
I have a folder with 44 images under `docs/images` (made of png, jgf, gif), weighting a total of 2.1MB.

Since I added that image folder, `mkdocs serve` takes a very long time to build (on startup only). I'd say 10-15 minutes, while it was taking only a couple of seconds before.

I don't find anything relevant in the doc, am I missing something?

**Comment by waylan** [comment]
Each time you do a build, all of those images need to be copied from the `docs_dir` to the `site_dir`. Naturally, the more there is to copy, the longer it will take. Although 10-15 minutes for 2.1MB seems a little excessive. 

Do you see the same behavior using the `mkdocs build` command? There is the `--dirty` option on `build` which only replaces updated files. That would make subsequent builds faster. However,  the `serve` command builds to a newly created temp dir, so that won't help on `serve`.

We'll need some more info about you setup to determine what is causing such long copy times.

**Comment by fclaerho** [comment]
Hello,
I'm observing the same issue with a directory of static resources. It takes two minutes for mkdocs to copy the files. It's not a massive dataset, I just have 33 pdf files at the moment.

Adding `--verbose` to the `build` command (`python2.7 -m mkdocs  build --strict --verbose --site-dir .build/site`), shows that it crawls to an halt at:
```
DEBUG   -  Copying static assets from the docs dir.
```

Using `--dirty` does help (the compilation is near instant in that case.)
I did not check the code, but that looks very weird that copying a few files takes so long, given they are not (supposed to be) processed in any way. Is it expected?

I'll keep using `--dirty` in the meantime.

**Comment by waylan** [comment]
What version of MkDocs are you using? 

The changes in #1294 were included in version 0.17. I'm curious to know if those changes introduced the problem or if it existed before that. Either way, we are just using Python's shutil standard lib file copying functions. We can't make a `cp -r` system call as we only copy non-markdown files.

**Comment by fclaerho** [comment]
```
❯ pip show mkdocs
Name: mkdocs
Version: 0.17.0
```

Also yes, I was not implying to use `cp -r` at all, it was just as a comparison to rule out some latency on my system.

**Comment by waylan** [comment]
Awesome! Thanks for doing the research on this. At least we now know the source of the problem and have a possible path forward. I'm out-of-town for the next week, but PRs are always welcome.

**Comment by fclaerho** [comment]
You're welcome!
Before starting to work on anything tho, that would be helpful is OP could confirm that using that hack above does solve his initial problem as well, just to be sure.

**Comment by sebw** [comment]
I will check tomorrow and report back.

**Comment by fclaerho** [comment]
Thanks.

The file is ~/Library/Python/2.7/lib/python/site-packages/mkdocs/utils/__init__.py (on macOS.)

And the part discussed above is:
```
102 def copy_file(source_path, output_path):
[…}
113     os.system("cp '{}' '{}'".format(source_path, output_path))
114     #shutil.copyfile(source_path, output_path)
```

**Comment by sebw** [comment]
I was running `0.17` at the time of opening the ticket.

I upgraded tonight:

```
# pip show mkdocs
Name: mkdocs
Version: 0.17.2
Summary: Project documentation with Markdown.
Home-page: http://www.mkdocs.org
Author: Tom Christie
Author-email: tom@tomchristie.com
License: BSD
Location: /usr/local/lib/python2.7/dist-packages
Requires: tornado, livereload, PyYAML, Jinja2, Markdown, click
```

When I run `mkdocs serve --verbose`, copying static assets is near instant, then it stays on:

```
DEBUG   -  Building page index.md
```

The situation is definitely better:

```
real	1m53.348s
user	1m50.230s
sys	0m2.160s
```
If I remove the images folder:

```
real	1m10.006s
user	1m6.540s
sys	0m2.100s
```
Is that considered a reasonable duration or is it still considered slow (I have 428 md files)?

**Comment by fclaerho** [comment]
Independently of the question if it's reasonable or still slow, I believe it's different from my "large-file" coy issue; I'll open another bug for my issue to not mix everything.

**Comment by waylan** [comment]
See this for a performance improvements to copying large files: http://blogs.blumetech.com/blumetechs-tech-blog/2011/05/faster-python-file-copy.html

**Comment by waylan** [comment]
After creating and testing #1534 it does not appear that there is really much improvement we can gain here. Especially when anything we would do would be nonstandard experimental code. I'm not comfortable including that in a production tool. If someone can point to an existing, well tested third party lib we can use, I'll reconsider.

**Comment by moerv9** [comment]
I've had the same problem, that `mkdocs serve` seemed to be stuck in the loading process.
I tried using `--dirtyreload` but it did not help.

With using `--verbose` I found out that the problem was inside one of my jupyter notebooks. 
In the mkdocs.yml file I've added that the Code should be executed before publishing. 
```
plugins:
  - search
  - mkdocs-jupyter:
      theme: dark
      execute: True
      include_source: True
```

When looking into the notebook there was an error in a cell and after fixing this everything worked!
